src/libraries/System.Text.Json/docs/KeyConverter_spec.md

```diff
@@ -0,0 +1,185 @@
+# Extend support for TKey in Dictionaries to non-string types
+
+# Motivation
+Most of our users that serialize dictionary use `Dictionary<string, TKey>`; however there is a significant amount that relies on `Dictionary<TKey, TValue>` where `TKey` is a primitive other than `string`, e.g: `int` or `Guid`, most of them came form `Newtonsoft.Json` which offers a plenty amount of support for using several types as the `TKey`, other popular .Net serializers also offer support for other types, the most common are integers (`int`, `uint`, `long`, etc.), and `enum`s (including Flags `enum`s).
+
+# Goals
+* 80%+ of dictionaries with non-string keys work out of the box, especially if they can round-trip.
+* Remain high performance.
+
+# Non-goals 
+* Complete parity with `Newtonsoft.Json` capabilities, especially in how string support is extended; any extension point can be through `JsonConverter<MyDictionary<non-string, TValue>>`.
+
+# Sample
+```cs
+// (De)serialize into a dictionary with a non-string key.
+Dictionary<int, string> root = new Dictionary<int, string>();
+root.Add(1, "value");
+
+string json = JsonSerializer.Serialize(root);
+// JSON
+// {
+//   "1":"value"
+// }
+
+Dictionary<int, string> rootCopy = JsonSerializer.Deserialize<Dictionary<int, string>>(json);
+Console.WriteLine(rootCopy[1]);
+ // Prints 
+ // value
+```
+
+# Strawman proposal
+
+## `KeyConverter`
+Implement an internal custom mechanism that is in charge of converting a defined set of types to be supported as the dictionary `TKey`; more or less like internal `JsonConverter`s work but for dictionary keys to JSON property names and viceversa.
+
+* The alternative that offers the best performance.
+
+* We need to define a criteria to choose what types we should support, I suggest to do as Utf8JsonReader/Writer and support the types supported by the Utf8Parser/Formatter.
+
+* Supported types (Types supported by [`Utf8Formatter/Parser`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.text.utf8formatter?view=netcore-3.1) + a few others that are popular):
+  * Boolean
+  * Byte
+  * DateTime
+  * DateTimeOffset
+  * Decimal
+  * Double
+  * Enum
+  * Guid
+  * Int16
+  * Int32
+  * Int64
+  * Object (Only on Serialization and if the runtime type is one of the supported types in this list)
+  * SByte
+  * Single
+  * String
+  * UInt16
+  * UInt32
+  * UInt64
+
+* https://github.com/Jozkee/runtime/tree/TKey_CustomConverter
+
+# Alternative considered
+
+## `TypeConverter`
+Use `TypeConverter` to parse and write the string representation of the type and use that as the JSON property name.
+* Many .NET types count with `TypeConverter` support that provides a string interpretation of the type.
+* Users can hook up `TypeConverter`s for their own types.
+* Not a high-performance alternative, involves unnecessary boxing.
+* https://github.com/Jozkee/runtime/tree/TKeySupport_TypeConverter
+
```

Comment:
I suppose an alternative is to not use string property names and instead use Name\Value combinations. Name\Value would likely be faster an just a few lines of code to enable since we don't need to add any new converter support.